library/Booster/Pattern/Rewrite.hs

```diff
@@ -128,17 +128,89 @@ rewriteStep cutLabels terminalLabels pat = do
             uniqueId = (.uniqueId) . (.attributes)
 
         case results of
+            -- no rules in this group were applicable
             [] ->
                 processGroups rest
-            [(r, x)]
+            -- there was a single rule which applied
+            [Applied (r, x)]
                 | labelOf r `elem` cutLabels ->
                     pure $ RewriteCutPoint (labelOf r) (uniqueId r) pat x
                 | labelOf r `elem` terminalLabels ->
                     pure $ RewriteTerminal (labelOf r) (uniqueId r) x
                 | otherwise ->
                     pure $ RewriteFinished (Just $ ruleLabelOrLocT r) (uniqueId r) x
-            rxs ->
-                pure $ RewriteBranch pat $ NE.fromList $ map (\(r, p) -> (labelOf r, uniqueId r, p)) rxs
+            rxs
+                -- all rules which did apply had an ensures condition which evaluated to false
+                | all (== Trivial) rxs ->
+                    -- if, all the other groups only generate a not applicable or trivial rewrites,
+                    -- then we return a `RewriteTrivial`.
+                    processGroups rest >>= \case
+                        RewriteStuck{} -> pure $ RewriteTrivial pat
+                        other -> pure other
+                -- at this point, there were some Applied rules and potentially some Trivial ones.
+                -- here, we just return all the applied rules in a `RewriteBranch`
+                | otherwise ->
+                    pure $
+                        RewriteBranch pat $
+                            NE.fromList $
+                                map (\(r, p) -> (labelOf r, uniqueId r, p)) $
+                                    concatMap (\case Applied x -> [x]; _ -> []) rxs
```

Comment:
Can this lead to pathological results where we return `RewriteBranch` but only have a single next state, e.g. for `results == [Applied something, Trivial, Trivial]`?
Maybe better to partition the `results` beforehand ? 
```
    let labelOf = ...
        (trivials, applieds) = partition (== Trivial) results
    if null results 
        then processGroups rest 
        else case applieds of
                [] -> ... -- all results were trivial for this group, proceed
                [Applied (r, x)] -> ... -- single result
                otherwise -> pure $ RewriteBranch pat applied
```

Reply:
Ah, good catch!